components/psst/psst_rule.cc

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/psst/psst_rule.h"
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/files/file_path.h"
+#include "base/json/json_reader.h"
+#include "base/types/expected.h"
+#include "components/prefs/pref_service.h"
+#include "extensions/common/url_pattern.h"
+#include "net/base/url_util.h"
+#include "url/gurl.h"
+#include "url/origin.h"
+#include "url/url_constants.h"
+
+namespace {
+
+// Format:
+// [
+//   {
+//     "include": [
+//       "*://twitter.com"
+//     ],
+//     "exclude": [
+//     ],
+//     "version": 1,
+//     "test_script": "twitter/test.js",
+//     "policy_script": "twitter/policy.js"
+//   }, ...
+// ]
+// Note that "test_script" and "policy_script" give paths
+// relative to the component under scripts/
+
+// psst.json keys
+const char kInclude[] = "include";
+const char kExclude[] = "exclude";
+const char kTestScript[] = "test_script";
+const char kPolicyScript[] = "policy_script";
+const char kVersion[] = "version";
+
+// Removes trailing dot from |host_piece| if any.
+// Copied from extensions/common/url_pattern.cc
+base::StringPiece CanonicalizeHostForMatching(base::StringPiece host_piece) {
+  if (base::EndsWith(host_piece, ".")) {
+    host_piece.remove_suffix(1);
+  }
+  return host_piece;
+}
+
+}  // namespace
+
+namespace psst {
+
+PsstRule::PsstRule() = default;
+
+PsstRule::~PsstRule() = default;
+
+// static
+bool PsstRule::GetURLPatternSetFromValue(const base::Value* value,
+                                         extensions::URLPatternSet* result) {
+  if (!value->is_list()) {
+    return false;
+  }
+  std::string error;
+  bool valid = result->Populate(
```

Comment:
why aren't we just defining these in a way that allows us to call `UrlPattern::Parse`?

Reply:
We need a URLPatternSet, not just one